CMakeLists.txt

```diff
@@ -0,0 +1,308 @@
+cmake_minimum_required(VERSION 3.12)
+project(libgambatte CXX)
+
+if(CMAKE_SOURCE_DIR STREQUAL PROJECT_SOURCE_DIR)
+  set(LIBGAMBATTE_IS_SUBPROJECT FALSE)
+else()
+  set(LIBGAMBATTE_IS_SUBPROJECT TRUE)
+endif()
+
+include(GNUInstallDirs)
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")
+
+option(LIBGAMBATTE_BUILD_SHARED "Build the shared version of libgambatte" ON)
+option(LIBGAMBATTE_BUILD_STATIC "Build the static version of libgambatte" ON)
+option(LIBGAMBATTE_ENABLE_ZIP "Build zip support using zlib" ON)
+option(LIBGAMBATTE_ENABLE_TESTING "Build tests" ON)
+set(LIBGAMBATTE_INSTALL_CMAKEDIR "${CMAKE_INSTALL_LIBDIR}/cmake/libgambatte" CACHE STRING "Directory to install the CMake package files to")
+set(LIBGAMBATTE_INSTALL_PKGCONFDIR "${CMAKE_INSTALL_LIBDIR}/pkgconfig" CACHE STRING "Directory to install the pkg-config file to")
+
+if(NOT LIBGAMBATTE_BUILD_SHARED AND NOT LIBGAMBATTE_BUILD_STATIC)
+  message(FATAL_ERROR "Both shared and static libraries have been disabled. Turn on at least one.")
+endif()
+
+if(NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE "Release" CACHE STRING "Build type" FORCE)
+  message(STATUS "No build type was specified, defaulting to: ${CMAKE_BUILD_TYPE}")
+  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "Debug;Release;MinSizeRel;RelWithDebInfo")
+endif()
+
+include(GetAvailableCompileFlags)
+get_available_compile_flags(LIBGAMBATTE_COMPILE_FLAGS)
+
+set(LIBGAMBATTE_SOURCES
+  "libgambatte/src/bitmap_font.cpp"
+  "libgambatte/src/cpu.cpp"
+  "libgambatte/src/gambatte.cpp"
+  "libgambatte/src/initstate.cpp"
+  "libgambatte/src/interrupter.cpp"
+  "libgambatte/src/interruptrequester.cpp"
+  "libgambatte/src/loadres.cpp"
+  "libgambatte/src/memory.cpp"
+  "libgambatte/src/newstate.cpp"
+  "libgambatte/src/sound.cpp"
+  "libgambatte/src/state_osd_elements.cpp"
+  "libgambatte/src/statesaver.cpp"
+  "libgambatte/src/tima.cpp"
+  "libgambatte/src/video.cpp"
+  "libgambatte/src/mem/camera.cpp"
+  "libgambatte/src/mem/cartridge.cpp"
+  "libgambatte/src/mem/huc3_chip.cpp"
+  "libgambatte/src/mem/infrared.cpp"
+  "libgambatte/src/mem/memptrs.cpp"
+  "libgambatte/src/mem/pakinfo.cpp"
+  "libgambatte/src/mem/remote.cpp"
+  "libgambatte/src/mem/rtc.cpp"
+  "libgambatte/src/mem/sgb.cpp"
+  "libgambatte/src/mem/time.cpp"
+  "libgambatte/src/mem/mbc/huc1.cpp"
+  "libgambatte/src/mem/mbc/huc3.cpp"
+  "libgambatte/src/mem/mbc/m161.cpp"
+  "libgambatte/src/mem/mbc/mbc0.cpp"
+  "libgambatte/src/mem/mbc/mbc1.cpp"
+  "libgambatte/src/mem/mbc/mbc2.cpp"
+  "libgambatte/src/mem/mbc/mbc3.cpp"
+  "libgambatte/src/mem/mbc/mbc5.cpp"
+  "libgambatte/src/mem/mbc/mmm01.cpp"
+  "libgambatte/src/mem/mbc/pocket_camera.cpp"
+  "libgambatte/src/mem/mbc/wisdom_tree.cpp"
+  "libgambatte/src/mem/snes_spc/dsp.cpp"
```

Comment:
Perhaps we should consider making snes_spc a separate "library" in this regard instead of bundled into libgambatte (not really any different in practice since it'd always just a static library included in libgambatte). Although that's probably suited for a later time.